controllers/alertsconfig_controller.go

```diff
@@ -50,14 +71,212 @@ type AlertsConfigReconciler struct {
+		}
+		return ctrl.Result{Requeue: requeueFlag}, nil
+	}
+
+	//First time use case
+	if !utils.ContainsString(alertsConfig.ObjectMeta.Finalizers, alertsConfigFinalizerName) {
+		log.Info("New alerts config resource. Adding the finalizer", "finalizer", alertsConfigFinalizerName)
+
+		alertsConfig.ObjectMeta.Finalizers = append(alertsConfig.ObjectMeta.Finalizers, alertsConfigFinalizerName)
+		r.CommonClient.UpdateMeta(ctx, &alertsConfig)
+		//That's fine- Let it come for requeue and we can create the alert
+		return ctrl.Result{}, nil
+	}
+
+	// Convert list of alerts from status into Map for easy access since
+	// we need to compare alert hash from status to newly calculated hash to find the difference
+	alertHashMap := alertsConfig.Status.Alerts
+
+	// if len(configs) > 0
+	for _, config := range alertsConfig.Spec.Alerts {
+		// Calculate checksum and compare it with the status checksum
+		exist, reqChecksum := utils.CalculateAlertConfigChecksum(ctx, config)
+		// if request and status checksum matches then there is change in this specific alert config
+		if exist && alertHashMap[config.AlertName].LastChangeChecksum == reqChecksum {
+			log.Info("checksum is equal. skipping")
+			//skip it
+			continue
+		}
+		// if there is a diff
+		// Get Alert CR
+
+		var wfAlert alertmanagerv1alpha1.WavefrontAlert
+		wfAlertNamespacedName := types.NamespacedName{Namespace: req.Namespace, Name: config.AlertName}
+		if err := r.Get(ctx, wfAlertNamespacedName, &wfAlert); err != nil {
+			log.Error(err, "unable to get the wavefront alert details for the requested name", "wfAlertName", config.AlertName)
+
+			// Update the status and retry it
+			return r.PatchIndividualAlertsConfigError(ctx, &alertsConfig, config.AlertName, err)
+		}
+		wfAlertBytes, err := json.Marshal(wfAlert.Spec)
+		if err != nil {
+			// update the status and retry it
+			return r.PatchIndividualAlertsConfigError(ctx, &alertsConfig, config.AlertName, err)
+		}
+
+		// execute Golang Template
+		wfAlertTemplate, err := template.ProcessTemplate(ctx, string(wfAlertBytes), config.Params)
+		if err != nil {
+			//update the status and retry it
+			return r.PatchIndividualAlertsConfigError(ctx, &alertsConfig, config.AlertName, err)
+		}
+		log.Info("Template process is successful", "here", wfAlertTemplate)
+
+		// Unmarshal back to wavefront alert
+		if err := json.Unmarshal([]byte(wfAlertTemplate), &wfAlert.Spec); err != nil {
+			// update the wfAlert status and retry it
+			return r.PatchIndividualAlertsConfigError(ctx, &alertsConfig, config.AlertName, err)
+		}
+		// Convert to Alert
+		var alert wf.Alert
+
+		if err := wavefront.ConvertAlertCRToWavefrontRequest(ctx, wfAlert.Spec, &alert); err != nil {
+			errMsg := "unable to convert the wavefront spec to Alert API request. will not be retried"
+			log.Error(err, errMsg)
+			return r.PatchIndividualAlertsConfigError(ctx, &alertsConfig, config.AlertName, err)
+		}
+
+		// Create/Update Alert
+		if alertHashMap[config.AlertName].ID == "" {
```

Comment:
Does alertHashMap have this alert for first create use case?

Reply:
No but if you are worried about nil case- since its a map it will not create any issue. it just gives us the zeroth value or empty

Reply:
if it is the first time creation use case the alertHashMap does not have the alert name, it will go through else{} which you have is update case, I think we should check for nil case in the same condition since it is a create case

Reply:
Hmmm - If the key value doesn't exist in the map (alertHashMap does not have the alert name case) the return value is the empty. Note that I'm not using pointer instead map[string]AlertStatus and I just tested it and it is working as expected